csrc/aes_cbc.cpp

```diff
@@ -172,96 +164,55 @@ class AesCbcCipher {
-
-        // The rest of the bytes for processing must come from input:
-        unsigned int const process_bytes_input = process_bytes - process_bytes_last_block;
-
-        // We also need to find out how many bytes should be saved from last_block and input by copying them into
-        // last_block. The total is save_bytes and we need to start saving from the tail of input.
-        unsigned int const save_bytes_input = min(save_bytes, input_len);
-        // The rest of the bytes that must be saved should come from the last_block itself.
-        unsigned int const save_bytes_last_block = save_bytes - save_bytes_input;
-
-        /* Let's prove two important theorems:
-        Theorem_1: process_bytes_last_block + save_bytes_last_block == last_block_len
-        Proof: expand each term to its defining expression:
-          min(process_bytes, last_block_len) + save_bytes - save_bytes_input ==
-          min(all - save_bytes, last_block_len) + save_bytes - min(save_bytes, input_len) ==
-          min(last_block_len + input_len - save_bytes, last_block_len) + save_bytes - min(save_bytes, input_len)
-
-        There are two possibilities:
-        1) save_bytes >= input_len: in this case:
-            min(last_block_len + input_len - save_bytes, last_block_len) == last_block_len &&
-            min(save_bytes, input_len) == save_bytes
-            Therefore, we can simplify the above equation to the following:
-            last_block_len + save_bytes - save_bytes  == last_block_len
-        2) save_bytes < input_len: in this case:
-            min(last_block_len + input_len - save_bytes, last_block_len) == last_block_len + input_len - save_bytes &&
-            min(save_bytes, input_len) == input_len
-            Therefore, we can simplify the above equation to the following:
-            last_block_len + input_len - save_bytes + save_bytes - input_len == last_block_len
-        */
-
-        /*
-        Theorem_2:  process_bytes_input + save_bytes_input == input_len
-        Proof: it can be proved similar to Theorem_1.
-        */
-
-        // Processing step: passing bytes to the cipher. This includes passing bytes from last_block and input:
-        // Passing bytes from last_block:
-        int result = update(last_block.get(), process_bytes_last_block, output, 0);
-
-        // Passing bytes from input
-        result += update(input, process_bytes_input, output + result, process_bytes_last_block - result);
-
-        // Copying step: storing bytes into last_byte. This includes copying from last_block into itself and from input
-        // to last_block.
-
-        // Copy save_bytes_last_block bytes from last_block to itself.
-        unsigned int from_index = last_block_len - save_bytes_last_block;
-        for (unsigned int i = 0; i < save_bytes_last_block; i++) {
-            last_block.get()[i] = last_block.get()[from_index + i];
+        if (all <= AES_CBC_BLOCK_SIZE_IN_BYTES) {
```

Comment:
This latest iteration is a lot easier to follow. I'm still not a fan of having to hold onto a full-block on each update.  This is not a problem for the logic in any other block algorithm. This seems to be caused by not having the right flags/context in between invocations, but I understand that this is a pretty annoying refactor to address.

One thing  you can do to simplify is to avoid special casing the `all=16b` case by just letting it flow through with a 0-byte update. That way we have one fewer edge case to track.

Reply:
After an offline convo, Amir walked me through why we need to manage this last full block on the decryption path.